File: MPP/LabC#/WindowsFormsApp1/repository/ProbaRepository.cs
using Concurs.model;
using Concurs.repository.utils;
using log4net;
using System;
using System.Collections.Generic;
using System.Data;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Concurs.repository
{
    public class ProbaRepository : IRepositoryProba
    {
        private static readonly ILog log = LogManager.GetLogger("ProbaRepository");

        IDictionary<String, string> props;
        public ProbaRepository(IDictionary<String, string> props)
        {
            log.Info("Creating ProbaRepository... ");
            this.props = props;
        }

        public Proba FindOne(int id)
        {
            log.InfoFormat("Entering findOne with value {0}", id);
            IDbConnection con = DBUtils.getConnection(props);

            using (var comm = con.CreateCommand())
            {
                comm.CommandText = "select id, denumire, categorie from Probe where id=@id";
                IDbDataParameter paramId = comm.CreateParameter();
                paramId.ParameterName = "@id";
                paramId.Value = id;
                comm.Parameters.Add(paramId);

                using (var dataR = comm.ExecuteReader())
                {
                    if (dataR.Read())
                    {
                        int idP = dataR.GetInt32(0);
                        string denumire = dataR.GetString(1);
                        string categorie = dataR.GetString(2);
                        Proba proba = new Proba(idP, denumire, categorie);
                        log.InfoFormat("Exiting findOne with value {0}", proba);
                        return proba;
                    }
                }
            }
            log.InfoFormat("Exiting findOne with value {0}", null);
            return null;
        }

        public IEnumerable<Proba> FindAll()
        {
            IDbConnection con = DBUtils.getConnection(props);
            log.InfoFormat("Entering FindAll...");
            IList<Proba> probe = new List<Proba>();
            using (var comm = con.CreateCommand())
            {
                comm.CommandText = "select id, denumire, categorie from Probe";

                using (var dataR = comm.ExecuteReader())
                {
                    while (dataR.Read())
                    {
                        int idP = dataR.GetInt32(0);
                        string denumire = dataR.GetString(1);
                        string categorie = dataR.GetString(2);
                        Proba proba = new Proba(idP, denumire, categorie);
                        probe.Add(proba);
                    }
                }
            }
            log.InfoFormat("Exiting findAll");
            return probe;
        }
        public void Save(Proba entity)
        {
            log.InfoFormat("Entering Save with new value {0}...", entity);
            var con = DBUtils.getConnection(props);
            using (var comm = con.CreateCommand())
            {
                comm.CommandText = "insert into Probe(denumire,categorie) values (@denumire, @categorie)";

                var paramDenumire = comm.CreateParameter();
                paramDenumire.ParameterName = "@denumire";
                paramDenumire.Value = entity.Denumire;
                comm.Parameters.Add(paramDenumire);

                var paramCategorie = comm.CreateParameter();
                paramCategorie.ParameterName = "@categorie";
                paramCategorie.Value = entity.Categorie;
                comm.Parameters.Add(paramCategorie);

                var result = comm.ExecuteNonQuery();
                if (result == 0)
                    throw new RepositoryException("Error: Nu s-a putut adauga proba!");
                log.InfoFormat("Exiting Save");
            }

        }
        public void Delete(int id)
        {
            log.InfoFormat("Entering Delete with new value {0}...", id);
            IDbConnection con = DBUtils.getConnection(props);
            using (var comm = con.CreateCommand())
            {
                comm.CommandText = "delete from Probe where id=@id";
                IDbDataParameter paramId = comm.CreateParameter();
                paramId.ParameterName = "@id";
                paramId.Value = id;
                comm.Parameters.Add(paramId);
                var rez = comm.ExecuteNonQuery();
                if (rez == 0)
                    throw new RepositoryException("Error: Proba nu s-a putut sterge!");
                log.InfoFormat("Exiting Delete");
            }
        }

        public void Update(int id, Proba entity)
        {
            log.InfoFormat("Entering Update with value {0}", id);
            var con = DBUtils.getConnection(props);

            using (var comm = con.CreateCommand())
            {
                comm.CommandText = "update Probe set  denumire=@denumire, categorie=@categorie where id=@id";
                var paramId = comm.CreateParameter();
                paramId.ParameterName = "@id";
                paramId.Value = id;
                comm.Parameters.Add(paramId);

                var paramDenumire = comm.CreateParameter();
                paramDenumire.ParameterName = "@denumire";
                paramDenumire.Value = entity.Denumire;
                comm.Parameters.Add(paramDenumire);

                var paramCategorie = comm.CreateParameter();
                paramCategorie.ParameterName = "@categorie";
                paramCategorie.Value = entity.Categorie;
                comm.Parameters.Add(paramCategorie);

                var result = comm.ExecuteNonQuery();
                if (result == 0)
                    throw new RepositoryException("Error: Nu s-a putut actualiza proba!");
                log.InfoFormat("Exiting Update");
            }
        }

        public int Size()
        {
            log.InfoFormat("Calculating database size...");
            IDbConnection con = DBUtils.getConnection(props);
            int size = -1;
            using (var comm = con.CreateCommand())
            {
                comm.CommandText = "select count(*) from Probe";

                var rez = comm.ExecuteNonQuery();
                if (rez == 0)
                    throw new RepositoryException("Size error !");

                using (var dataR = comm.ExecuteReader())
                {
                    if (dataR.Read())
                    {
                        size = dataR.GetInt32(0);
                    }
                }
            }
            log.InfoFormat("Finished calculation");
            return size;
        }

        public IEnumerable<string> ListaCategorii()
        {
            IDbConnection con = DBUtils.getConnection(props);
            log.InfoFormat("Entering ListaCategorii...");
            IList<string> categorii = new List<string>();
            using (var comm = con.CreateCommand())
            {
                comm.CommandText = "select distinct categorie from Probe";

                using (var dataR = comm.ExecuteReader())
                {
                    while (dataR.Read())
                    {
                        string categorie = dataR.GetString(0);
                        categorii.Add(categorie);
                    }
                }
            }
            log.InfoFormat("Exiting ListaCategorii");
            return categorii;
        }

        public IEnumerable<string> ListaProbeNume()
        {
            IDbConnection con = DBUtils.getConnection(props);
            log.InfoFormat("Entering ListaProbeNume...");
            IList<string> denumiri = new List<string>();
            using (var comm = con.CreateCommand())
            {
                comm.CommandText = "select distinct denumire from Probe";

                using (var dataR = comm.ExecuteReader())
                {
                    while (dataR.Read())
                    {
                        string denum = dataR.GetString(0);
                        denumiri.Add(denum);
                    }
                }
            }
            log.InfoFormat("Exiting ListaCategorii");
            return denumiri;
        }
    }
}

File: MPP/ClientServer_Java/Persistence/src/main/java/repository/ProbaRepository.java
package repository;

import model.Categorie;
import model.Proba;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;
import java.util.Properties;

public class ProbaRepository implements IRepositoryProba{
    private JdbcUtils dbUtils;

    private static final Logger logger= LogManager.getLogger();
    public ProbaRepository(Properties props){
        logger.info("Initializing ProbaRepository with properties: {} ",props);
        dbUtils=new JdbcUtils(props);
    }
    @Override
    public int size() {
        logger.traceEntry();
        Connection con=dbUtils.getConnection();
        try(PreparedStatement preStmt=con.prepareStatement("select count(*) as [SIZE] from Probe")) {
            try(ResultSet result = preStmt.executeQuery()) {
                if (result.next()) {
                    logger.traceExit(result.getInt("SIZE"));
                    return result.getInt("SIZE");
                }
            }
        }catch(SQLException ex){
            logger.error(ex);
            System.out.println("Error DB "+ex);
        }
        return 0;
    }

    @Override
    public Integer save(Proba entity) {
        logger.traceEntry("saving proba {} ",entity);
        Connection con=dbUtils.getConnection();
        try(PreparedStatement preStmt=con.prepareStatement("insert into Probe(denumire, categorie) values (?,?)")){
            preStmt.setString(1,entity.getDenumire());
            preStmt.setString(2,entity.getCatg().toString());
            int result=preStmt.executeUpdate();
            if(result==0) throw new RepositoryException("Error: Nu s-a putut salva proba!");
            try(PreparedStatement preStmt1=con.prepareStatement("select top 1 id from Probe order by id desc")) {
                try(ResultSet result1 = preStmt.executeQuery()) {
                    if (result1.next()) {
                        return result1.getInt("id");
                    }
                }
            }
        }catch (SQLException ex){
            logger.error(ex);
            System.out.println("Error DB "+ex);
        }
        logger.traceExit();
        return null;
    }

    @Override
    public void delete(Integer integer) {
        logger.traceEntry("deleting proba with {}",integer);
        Connection con=dbUtils.getConnection();
        try(PreparedStatement preStmt=con.prepareStatement("delete from Probe where id=?")){
            preStmt.setInt(1,integer);
            int result=preStmt.executeUpdate();
            if(result==0) throw new RepositoryException("Error: Nu s-a putut sterge proba!");
        }catch (SQLException ex){
            logger.error(ex);
            System.out.println("Error DB "+ex);
        }
        logger.traceExit();
    }

    @Override
    public void update(Integer integer, Proba entity) {
        logger.traceEntry("updating proba with {}",integer);
        Connection con=dbUtils.getConnection();

        try(PreparedStatement preStmt=con.prepareStatement("update Probe " +
                "set denumire=?, categorie=?" +
                "where id=?")){
            preStmt.setInt(3,integer);
            preStmt.setString(1,entity.getDenumire());
            preStmt.setString(2,entity.getCatg().toString());
            int result=preStmt.executeUpdate();
            if(result==0) throw new RepositoryException("Error: Nu s-a putut actualiza proba!");
        }catch (SQLException ex){
            logger.error(ex);
            System.out.println("Error DB "+ex);
        }
        logger.traceExit();
    }

    @Override
    public Proba findOne(Integer integer) {
        logger.traceEntry("finding proba with id {} ",integer);
        Connection con=dbUtils.getConnection();

        try(PreparedStatement preStmt=con.prepareStatement("select * from Probe where id=?")){
            preStmt.setInt(1,integer);
            try(ResultSet result=preStmt.executeQuery()) {
                if (result.next()) {
                    Integer id = result.getInt("id");
                    String descriere = result.getString("denumire");
                    String categorie = result.getString("categorie");
                    Proba proba=new Proba(id,descriere,Categorie.valueOf(categorie));
                    logger.traceExit(proba);
                    return proba;
                }
            }
        }catch (SQLException ex){
            logger.error(ex);
            System.out.println("Error DB "+ex);
        }
        logger.traceExit("No proba found with id {}", integer);

        return null;
    }

    @Override
    public Iterable<Proba> findAll() {
        logger.traceEntry();
        Connection con=dbUtils.getConnection();
        List<Proba> probe=new ArrayList<>();
        try(PreparedStatement preStmt=con.prepareStatement("select * from Probe")) {
            try(ResultSet result=preStmt.executeQuery()) {
                while (result.next()) {
                    Integer id = result.getInt("id");
                    String descriere = result.getString("denumire");
                    String categorie = result.getString("categorie");
                    Proba proba=new Proba(id,descriere,Categorie.valueOf(categorie));
                    probe.add(proba);
                }
            }
        } catch (SQLException e) {
            logger.error(e);
            System.out.println("Error DB "+e);
        }
        logger.traceExit(probe);
        return probe;
    }

    @Override
    public Iterable<Categorie> listaCategorii() {
        logger.traceEntry();
        Connection con=dbUtils.getConnection();
        List<Categorie> categorii=new ArrayList<>();
        try(PreparedStatement preStmt=con.prepareStatement("select distinct categorie from Probe")) {
            try(ResultSet result=preStmt.executeQuery()) {
                while (result.next()) {
                    String categorie = result.getString("categorie");
                    categorii.add(Categorie.valueOf(categorie));
                }
            }
        } catch (SQLException e) {
            logger.error(e);
            System.out.println("Error DB "+e);
        }
        logger.traceExit(categorii);
        return categorii;
    }

    @Override
    public Iterable<String> listaProbeNume() {
        logger.traceEntry();
        Connection con=dbUtils.getConnection();
        List<String> probe=new ArrayList<>();
        try(PreparedStatement preStmt=con.prepareStatement("select distinct denumire from Probe")) {
            try(ResultSet result=preStmt.executeQuery()) {
                while (result.next()) {
                    String denumire = result.getString("denumire");
                    probe.add(denumire);
                }
            }
        } catch (SQLException e) {
            logger.error(e);
            System.out.println("Error DB "+e);
        }
        logger.traceExit(probe);
        return probe;
    }
}
